app/assets/javascripts/services/alerts_center_service.js

```diff
@@ -0,0 +1,828 @@
+    },
+    {
+      id: 'addcomment',
+      name: __('Add Note'),
+      isVisible: true,
+      actionFn: handleMenuAction
+    },
+    {
+      id: 'assign',
+      name: __('Assign'),
+      isVisible: true,
+      actionFn: handleMenuAction
+    },
+    {
+      id: 'unacknowledge',
+      name: __('Unacknowledge'),
+      isVisible: true,
+      actionFn: handleMenuAction
+    },
+    {
+      id: 'unassign',
+      name: __('Unassign'),
+      isVisible: true,
+      actionFn: handleMenuAction
+    }
+  ];
+
+  _this.updateMenuActionForItemFn = function(action, item) {
+    if (action.id === 'unassign') {
+      action.isVisible = item.assigned;
+    } else if (action.id === 'acknowledge') {
+      action.isVisible = (item.assignee_id == _this.currentUser.id) && item.acknowledged !== true;
+    } else if (action.id === 'unacknowledge') {
+      action.isVisible = (item.assignee_id == _this.currentUser.id) && item.acknowledged === true;
+    } else {
+      action.isVisbile = true;
+    }
+  };
+
+  _this.getUserByIdOrUserId = function(id) {
+    var foundUser;
+    for (var i = 0; i < _this.existingUsers.length && !foundUser; i++) {
+      if (_this.existingUsers[i].id === id || _this.existingUsers[i].userid === id) {
+        foundUser = _this.existingUsers[i];
+      }
+    }
+
+    return foundUser;
+  };
+
+  _this.getCurrentUser = function() {
+    var deferred = $q.defer();
+
+    // Get the current user
+    API.get('/api').then(function(response) {
+      _this.currentUser = response.identity;
+      deferred.resolve();
+    });
+
+    return deferred.promise;
+  };
+
+  _this.updateExistingUsers = function() {
+    var deferred = $q.defer();
+
+    // Get the existing users
+    API.get('/api/users?expand=resources').then(function (response) {
+      // update the existing users list and current user
+      _this.existingUsers = response.resources;
+      _this.existingUsers.sort(function(user1, user2) {
+        return user1.name.localeCompare(user2.name);
+      });
+      _this.currentUser = _this.getUserByIdOrUserId(_this.currentUser.userid);
+
+      deferred.resolve();
+    });
+
+    return deferred.promise;
+  };
+
+  _this.updateProviders = function() {
+    var deferred = $q.defer();
+
+    API.get(providersURL + '?expand=resources&attributes=tags').then(function(response) {
+      _this.providers = response.resources;
+      deferred.resolve();
+    });
+
+    return deferred.promise;
+  };
+
+  _this.updateTags = function() {
+    var deferred = $q.defer();
+
+    API.get(tagsURL + '?expand=resources&attributes=category,categorization').then(function(response) {
+      _this.tags = response.resources;
+      deferred.resolve();
+    });
+
+    return deferred.promise;
+  };
+
+  _this.updateAlertsData = function(limit, offset, filters, sortField, sortAscending) {
+    var deferred = $q.defer();
+
+    // Update data then get the alerts data
+    _this.getCurrentUser().then(function() {
```

Comment:
nesting hell is precisely the reason to use Promises ;)

this could be rewritten as..

```js
_this.updateAlertsData = function(limit, offset, filters, sortField, sortAscending) {
  return _this.getCurrentUser()
    .then(_this.updateExistingUsers)
    .then(_this.updateProviders)
    .then(_this.updateTags)
    .then(function() {
      return _this.getAlertsData(limit, offset, filters, sortField, sortAscending);
    });
};
```